### PR TITLE
Rework client-pool-timer-test to avoid races

### PR DESCRIPTION
**Pull request by upshaw-alex**
This PR reworks `client-pool-timer-test` (a unit test suite implemented in `client/client_pool/test/client_pool_timer_test.cpp`) to avoid race conditions the existing implementation contains, which mostly arose from the existing implementation relying on the timing of timeout callbacks made by the `concord_client_pool::Timer` being tested in ways the Timer does not really guarantee.

This PR also reactivates this test suite, as it was previously deactivated due to its instability. This PR additionally adds some comments to the header file for `concord_client_pool::Timer` (`client/client_pool/include/client_pool/client_pool_timer.hpp`) to try to clarify its expected behavior.

* **Problem Overview**  
Intermittent failures of `client-pool-timer-test` were observed. Specifically, testing assertions in its [existing `work_items` test case](https://github.com/vmware/concord-bft/blob/e9488e6d44db412dd568f1f88e3fe8d19c07c75c/client/client_pool/test/client_pool_timer_test.cpp#L26) could fail. The test case's approach was to construct a `concord_client_pool::Timer` with a timeout callback function that increments a counter and then test that it made timeout callbacks as expected by calling its `start` method, waiting twice the `Timer`'s timeout duration, then expecting the counter to have been incremented. However, expecting the callback to have completed asynchronously within a fixed time frame is inherently not completely reliable, given this is not a hard real-time system and the scheduler could end up delaying and/or interrupting the asynchronous timeout callback. Furthermore, the existing test tested that `concord_client_pool::Timer`'s `cancel` and `stopTimerThread` actually stopped timeout callbacks by starting a timeout, immediately cancelling it, then checking that the counter was not incremented. This is also inherently not completely reliable since the timeout is racing its cancellation, and, for example, the callback could be made before it gets cancelled if the main thread running the unit test case gets interrupted or preempted between the time it starts the timer and the time it completes the cancellation.

* **Testing Done**  
With these changes, I have been able to run the `client-pool-timer-test` suite both with `DCMAKE_BUILD_TYPE=DEBUG` and with `DCMAKE_BUILD_TYPE=Release` 256 consecutive times each without seeing any failures.

### CODE REVIEW

**Comment by upshaw-alex** [comment]
For context, I might add: I chose to re-work this test suite somewhat rather than making the bare minimum modifications necessary to prevent the race conditions since I could not tell if the existing test suite was deliberately intended to actually test timing, and I was concerned that, if that was its intention, it might leave it in an incoherent state where it wasn't clear what properties/behaviors it was trying to test if it was written to test timing but I made it overlook timing because timing is unreliable.

I attempted to add more explicit documentation of what it looks like the intended behaviors of `concord_client_pool::Timer` are to its header file in order to better understand myself and document for the record what behaviors of this `Timer` class it is valid to test. At least ideally, as part of this review, I think we might want someone who considers themselves familiar with the intended behavior and usage of this `Timer` class to confirm my attempt to document its behavior is accurate and that I have not misinterpreted here.

**Comment by teoparvanov** [comment]
Hi @upshaw-alex , thanks for the clarification! I'd like to hear from @yuliasherman , @situ-s or @shgandhi before this gets approved, as the test is a sort of a specification for the tested component's behavior.

**Comment by upshaw-alex** [comment]
I have force pushed an amendment to my commit for this PR to address some warnings (treated as errors) that Clang-tidy was producing; I have added some `NOLINT` comments to suppress these warnings (treated as errors), which are from the `bugprone-infinite-loop` check, as I believe the loops it is identifying as infinite are not infinite. They do have loop conditions depending on only a local variable that does not get modified in the loop, but I expect the variable to be updated asynchronously by the `Timer` objects under test, which are constructed with lambda expressions including a capture-by-reference of the local variable the loops are waiting on.

I have also taken the opportunity to make `kTestTimeout` `const`, as I noticed I somehow neglected to do so in the process of creating this amendment when I was working on an inelegant workaround for the Clang-tidy warnings before it occurred to me I should check to see if Clang-tidy has a way to just manually suppress them.

**Comment by upshaw-alex** [comment]
I have force-pushed a rebase of my commit for this PR off of `master` in the interest of keeping this PR up-to-date.

File: client/client_pool/test/client_pool_timer_test.cpp [hunk 1]
```diff
@@ -19,50 +19,154 @@
...
+using std::make_shared;
+using std::make_unique;
+using std::shared_ptr;
+
 using namespace std::chrono_literals;
 
-using TestClient = std::shared_ptr<void>;
+using TestClient = shared_ptr<void>;
+
+const milliseconds kTestTimeout = 1ms;
```
Comment (yuliasherman) [inline client/client_pool/test/client_pool_timer_test.cpp:31]: The timeout feels too short. In the weak CI environment, it could cause instability. I would increase it up to 3-5 ms.

**Review by shgandhi — COMMENTED** [review]
Thanks for the change @upshaw-alex. LGTM. 
I have added a few minor comments.

File: client/client_pool/test/client_pool_timer_test.cpp [hunk 2]
```diff
@@ -19,50 +19,154 @@
...
-
-  // Starting new timer won't work because the thread is stopped
-  TestClient client6;
-  timer.start(client6);
-  std::this_thread::sleep_for(timeout * 2);
-  ASSERT_EQ(num_times_called, 3);
+  while (num_times_called < 3) {  // NOLINT(bugprone-infinite-loop): timer should update num_times_called
+    std::this_thread::sleep_for(kTestTimeout * 2);
+  }
+  EXPECT_EQ(num_times_called, 3) << "Timer::start caused an unexpected number of callbacks to timeout funciton.";
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp:62]: I'm not sure I follow your concern here.

> couldn't it happen that you get more callback calls than you expect here.
This test makes 3 total calls to `Timer::start`, and I do not think 1 call to `Timer::start` should cause more than 1 callback, so I do not think it should be possible to get more than 3 callbacks by the time of this check. Is there something I am missing here?

File: client/client_pool/test/client_pool_timer_test.cpp (hunk 1 above)
Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp:31]: I am fine with increasing the timeout to 3-5ms, though I am curious how you think this could could cause instability, as I believe I have tried to structure this reworked test suite not to rely on any unreliable timing. Do you have a specific example of how having a low value for `kTestTimeout` could cause these tests to fail?

File: client/client_pool/test/client_pool_timer_test.cpp [hunk 3]
```diff
@@ -19,50 +19,154 @@
...
-  // Cancel timeout
+  // Note while we expect the timer to make a call back after about kTestTimeout, it makes no hard timing guarantees,
+  // so we may have to wait longer sometimes.
+  while (num_times_called < 1) {  // NOLINT(bugprone-infinite-loop): timer should update num_times_called
+    std::this_thread::sleep_for(kTestTimeout * 2);
+  }
+  EXPECT_EQ(num_times_called, 1) << "Timer::start caused an unexpected number of callbacks to timeout funciton.";
+
+  std::this_thread::sleep_for(kTestTimeout * 20);
+  EXPECT_LE(num_times_called, 1) << "Timer::start appears to have caused extraneous callback(s) after the initial one.";
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp:48]: I originally chose `EXPECT_LE` here because the purpose of this specific check is to expect that the `Timer` implementation does not generate any additional extraneous callbacks to the timeout function, so I wanted to add an expectation that `num_times_called` is not greater than 1, and `EXPECT_LE` is equivalent to expecting greater than to be false. Now that you mention it however, I will plan to change this to `EXPECT_EQ`, as there is no valid reason `num_times_called` should be any lower than `1` either, and I guess using `EXPECT_LE` here kind of suggests the test is intentionally accounting for such a possibility.

File: client/client_pool/include/client/client_pool/client_pool_timer.hpp [hunk 4]
```diff
@@ -48,6 +70,16 @@ class Timer {
...
   }
 
+  /*
+   * Start the timer for a timeout, and make an asynchronous call to the on_timeout function (given at this Timer's
+   * construction) with client as a parameter after the timeout duration (also given at this Timer's construction).
+   * Note that, though this Timer's implementation should make a reasonable effort to make this callback after about
+   * the given timeout duration, no hard guarantees about the precise timing of this callback are provided. No callback
+   * will be made if this Timer was constructed with a timeout of 0 milliseconds or if stopTimerThread has previously
+   * been called for this Timer. Behavior is undefined if start is called at any time when there is an ongoing timeout
+   * from a previous call to start that has not either completed its callback or been successfully cancelled via
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/include/client/client_pool/client_pool_timer.hpp:80]: I would also be interested to hear @yuliasherman's take on this if she is familiar with the intended API here, as the comments I have added to `client_pool_timer.hpp` and the re-worked tests include some interpretations I am not completely certain about.

[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/include/client/client_pool/client_pool_timer.hpp:80]: I will say that, under my current unconfirmed interpretations, I think the API does allow for successive `start` calls if the callbacks from all previous `start` calls have either already been completed or have been _successfully_ cancelled with `cancel`. The case where I have interpreted behavior as undefined is specifically the one where there is a callback scheduled by a previous `start` call still in flight at the time a new `start` call begins. This includes both the case where `start` is called once again in a row without cancelling or waiting for the previous callback, and the case where the previous call was cancelled with `cancel`, but that cancellation failed.

I will mention I think it is non-ideal that, at least under my current interpretation, `Timer::cancel` can effectively fail silently and thereby leave the `Timer` in a state where subsequent calls to `start` have undefined behavior. I am not quite sure what the best way to rectify this would be, though I might be interested to examine that further and try to produce a follow-up PR to clean this up.

**Comment by upshaw-alex** [comment]
Thank you very much for your reviews. I have force-pushed an amendment to my commit for this PR to make some of the revisions suggested. I did also have some further questions and comments on some of the review comments; please see the replies I have posted for those.

File: client/client_pool/include/client/client_pool/client_pool_timer.hpp [hunk 5]
```diff
@@ -23,11 +23,22 @@
...
+ *
+ * Note oddity: ClientT has to implement `operator<<` (for logging purposes)
+ */
 template <typename ClientT>
 class Timer {
  public:
   using Clock = std::chrono::high_resolution_clock;
+
+  /*
+   * Constructor for a Timer<ClientT> to (once started) asynchronosouly call on_timeout after timeout milliseconds. A
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/include/client/client_pool/client_pool_timer.hpp]: This typo should be fixed in the latest revisions for this PR.

File: client/client_pool/test/client_pool_timer_test.cpp [hunk 6]
```diff
@@ -19,50 +19,154 @@
...
-  timer.start(client6);
-  std::this_thread::sleep_for(timeout * 2);
-  ASSERT_EQ(num_times_called, 3);
+  while (num_times_called < 3) {  // NOLINT(bugprone-infinite-loop): timer should update num_times_called
+    std::this_thread::sleep_for(kTestTimeout * 2);
+  }
+  EXPECT_EQ(num_times_called, 3) << "Timer::start caused an unexpected number of callbacks to timeout funciton.";
+}
+
+TEST(client_pool_timer, cancelReturnsNonnegativeDuration) {
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp]: I have made this change in the latest revisions for this PR.

File: client/client_pool/test/client_pool_timer_test.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp]: I have made this change in the latest revisions for this PR.

File: client/client_pool/test/client_pool_timer_test.cpp [hunk 7]
```diff
@@ -19,50 +19,154 @@
...
+  EXPECT_LE(num_times_called, 1) << "Timer::start appears to have caused extraneous callback(s) after the initial one.";
+
   TestClient client2;
   timer.start(client2);
-  timer.cancel();
-  ASSERT_EQ(num_times_called, 1);
+  while (num_times_called < 2) {  // NOLINT(bugprone-infinite-loop): timer should update num_times_called
+    std::this_thread::sleep_for(kTestTimeout * 2);
+  }
+  EXPECT_EQ(num_times_called, 2) << "Timer::start caused an unexpected number of callbacks to timeout funciton.";
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp]: This typo should be fixed in the latest revisions for this PR.

**Comment by upshaw-alex** [comment]
I see there has been a failure in the `skvbc_reconfiguration` test suite for this PR. I don't believe this PR should affect that suite, so I am going to assume this is a transient failure unrelated to this PR (feel free to correct me if you think otherwise).

File: client/client_pool/test/client_pool_timer_test.cpp (hunk 1 above)
Comment (yuliasherman) [reply client/client_pool/test/client_pool_timer_test.cpp:31]: I'm just trying to say that in a non-realtime system small timeouts do not work precisely, especially in CI, which runs on top of very weak systems.

**Review by yuliasherman — APPROVED** [review]
LGTM

**Comment by upshaw-alex** [comment]
I have force-pushed an amendment to my commit for this PR to adopt @yuliasherman's suggestion for `kTestTimeout`.

**Comment by upshaw-alex** [comment]
I have force-pushed a rebase of my commit for this PR off of `master` in the interest of keeping this PR up-to-date.

File: client/client_pool/test/client_pool_timer_test.cpp (hunk 1 above)
Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp]: As an update, I have gone ahead and changed the timeout to 5ms in recent revisions.

Comment (upshaw-alex) [reply client/client_pool/test/client_pool_timer_test.cpp]: I understand small timeouts do not work precisely in non-realtime systems. For context as to what I was asking, I believe I have structured the re-worked tests to avoid relying on precise timing at all (and in fact, I would attribute the flakiness of this test suite prior to this re-work to it relying on precise timing). I am adopting the high end of your suggested timeout since it sounds like it might be better practice, though I do not think having an unreliably short timeout here would cause any flakiness in the re-worked tests if I have written them correctly.

**Review by teoparvanov — APPROVED** [review]
LGTM, thanks Alex!

File: client/client_pool/include/client/client_pool/client_pool_timer.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuliasherman) [reply client/client_pool/include/client/client_pool/client_pool_timer.hpp:80]: @upshaw-alex I think the explanation is good enough. There is no way "to make sure that no successive start calls for the same client are made".

[the remark this answers is not in the document]
Comment (shgandhi) [reply client/client_pool/include/client/client_pool/client_pool_timer.hpp:80]: Thanks for clearing that up @yuliasherman!
